### PR TITLE
Fix UIDevice's identifierForVendor

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
This pull request also fixes an NSUUID bug.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
This pull request switches our publisher ID generation to generate a uuidv5 over the package's publisher ID as dictated by its code signature. Different applications from the same publisher will get the same value for `-[UIDevice identifierForVendor]`.

Check out the gory details inside!

File: Frameworks/Foundation/NSUUID.mm [hunk 1]
```diff
@@ -100,6 +100,13 @@ - (id)initWithUUIDBytes:(const void*)bytes {
             [self release];
             return nil;
         }
+
```
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSUUID.mm]: no tests for these?

**Comment by DHowett-MSFT** [comment]
@rajsesh-msft Check it out, there are a whooooole lot of awesome new tests and fixes!

File: Frameworks/Foundation/NSUUID.mm [hunk 2]
```diff
@@ -100,6 +100,13 @@ - (id)initWithUUIDBytes:(const void*)bytes {
...
+        // The reference platform specifies that a UUID is a uint8_t[16], but the Win32
+        // definition is 3 typed integers plus a uint8_t[6].
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUUID.mm]: nit: `8` not `6`

File: Frameworks/Foundation/NSUUID.mm [hunk 3]
```diff
@@ -159,7 +169,25 @@ - (NSString*)UUIDString {
...
+    buffer[2] = (_guid.Data1 >>  8) & 0xFF;
+    buffer[3] = (_guid.Data1      ) & 0xFF;
+
+    buffer[4] = (_guid.Data2 >>  8) & 0xFF;
+    buffer[5] = (_guid.Data2      ) & 0xFF;
+
+    buffer[6] = (_guid.Data3 >>  8) & 0xFF;
+    buffer[7] = (_guid.Data3      ) & 0xFF;
+
+    memcpy_s(&buffer[8], 8, &_guid.Data4, 8);
```
Comment (aballway) [inline Frameworks/Foundation/NSUUID.mm]: nit: buffer + 8

File: tests/unittests/Foundation/NSUUIDTests.mm [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    ASSERT_OBJCNE(nil, [NSUUID UUID]);
+}
+
+TEST(NSUUID, Uniqueness) {
+    ASSERT_OBJCNE([NSUUID UUID], [NSUUID UUID]);
+}
+
+TEST(NSUUID, CopyViaString) {
+    NSUUID* uuid = [NSUUID UUID];
+    StrongId<NSUUID> uuidClone{ woc::TakeOwnership, [[NSUUID alloc] initWithUUIDString:[uuid UUIDString]] };
```
Comment (aballway) [inline tests/unittests/Foundation/NSUUIDTests.mm]: super nit: just use autorelease for these

File: tests/unittests/Foundation/NSUUIDTests.mm [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+TEST(NSUUID, InitWithString) {
+    StrongId<NSUUID> uuid {
+        woc::TakeOwnership, [[NSUUID alloc] initWithUUIDString:@"01020304-0506-0708-090A-0B0C0D0E0F10"]
+    };
+
+    uint8_t bytes[16];
+    ASSERT_NO_THROW([uuid getUUIDBytes:(unsigned char*)bytes]);
+
+    std::vector<uint8_t> comparand(&bytes[0], bytes + 16);
```
Comment (aballway) [inline tests/unittests/Foundation/NSUUIDTests.mm]: nit: just `bytes`

File: tests/unittests/Foundation/NSUUIDTests.mm [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+TEST(NSUUID, UUIDString) {
+    StrongId<NSUUID> uuid {
+        woc::TakeOwnership, [[NSUUID alloc] initWithUUIDString:@"01020304-0506-0708-090A-0B0C0D0E0F10"]
+    };
+
+    ASSERT_OBJCEQ(@"01020304-0506-0708-090A-0B0C0D0E0F10", [uuid UUIDString]);
+}
+
+TEST(NSUUID, InvalidStrings) {
+    EXPECT_OBJCEQ(nil, [[NSUUID alloc] initWithUUIDString:@"HELLOWOR-LDTH-ISIS-ABAD-UUIDSTRING!!"]);
```
Comment (aballway) [inline tests/unittests/Foundation/NSUUIDTests.mm]: nit: no egyptian deities

File: Frameworks/Foundation/NSUUID.mm [hunk 7]
```diff
@@ -109,20 +116,23 @@ - (id)initWithUUIDBytes:(const void*)bytes {
...
+    if (self = [super init]) {
+        if ([coder isKindOfClass:[NSKeyedUnarchiver class]]) {
+            NSUInteger len = 0;
+            const uint8_t* bytes = [coder decodeBytesForKey:@"NSU.uuid" returnedLength:&len];
+            if (!bytes || len != sizeof(_guid)) {
+                [self release];
+                return nil;
+            }
+
+            memcpy_s(&_guid, 16, bytes, 16);
```
Comment (aballway) [inline Frameworks/Foundation/NSUUID.mm]: nit: magic numbers

**Pull request merged**